libgnucash/engine/Account.cpp

```diff
@@ -4392,6 +4392,79 @@ xaccAccountSetReconcileLastDate (Account *acc, time64 last_date)
     xaccAccountCommitEdit (acc);
 }
 
+
+void
+xaccAccountClearReconcileStatementData (Account *acc)
+{
+    if (!acc) return;
+
+    qof_instance_slot_path_delete
+        (QOF_INSTANCE(acc), {KEY_RECONCILE_INFO, "statements"});
+}
+
+GList *
+xaccAccountGetReconcileStatementDates (Account *acc)
+{
+    GValue v = G_VALUE_INIT;
+    GList *list = NULL;
+
+    if (!GNC_IS_ACCOUNT(acc)) return list;
+
+    // work in progress...
+    qof_instance_get_path_kvp
+        (QOF_INSTANCE(acc), &v, {KEY_RECONCILE_INFO, "statements"});
+
```

Comment:
The quick-and-dirty way is
```C++
auto slots = qof_instance_get_slots(QOF_INSTANCE(acc));
for (auto date : slots.get_slot({KEY_RECONCILE_INFO, "statements"}).get_keys())
    list = g_list_prepend(static_cast<time64>(GDateTime(date)));
list = g_list_sort(list, intsort); //Just reverse the list if the dates are kept sorted in the KVP frame
```
and intsort is easy
```C
int
intsort(gconstpointer pa, gconstpointer pb)
{
    int a = G_POINTER_TO_INT(pa);
    int b = G_POINTER_TO_INT(pb);
    return (a > b ? 1 : a < b ? -1 : 0);
}
```
But that breaks the encapsulation of KVP in qof_instance so it would be better to move the actual iteration into a qof_instance function similar to the `qof_instance_kvp_foo_guid` set.

Reply:
Now that looks like some C++ that I would never be able to do. It would eventually be used by `window-reconcile.c` to populate a drop-down or something to retrieve previous statement dates. The dropdown would also allow a `Delete` button.

A beefed-up `reconcile.scm` could also scan through all statement dates, and show a section whereby the reconciled balance do not match, thereby making transaction hunting easier.

I'd think the list must be sorted when retrieved because my vision is that we should always be able to grab an old random bank statement and reconcile it, thus adding a soft 'reconciled-balance assertion'.

Reply:
Ok the quick-fixed-and-safe version is
```c++
GList *
xaccAccountGetReconcileStatementDates (Account *acc)
{
    GList *list = nullptr;

    if (!GNC_IS_ACCOUNT(acc)) return list;

    auto slots = qof_instance_get_slots(QOF_INSTANCE(acc));
    if (!slots) return list;

    auto slot = slots->get_slot({KEY_RECONCILE_INFO, "statements"});
    if (!slot) return list;

    auto keys = slot.get_keys ();
    if (!keys) return list;

    for (auto datestr : keys)
    {
        auto date = gnc_iso8601_to_time64_gmt(datestr);
        if (date != INT64_MAX)
            list = g_list_prepend(date);
    }

    list = g_list_sort(list, intsort);

    return list;
}
```

But I can't fix gcc error:
```
FAILED: libgnucash/engine/CMakeFiles/gncmod-engine.dir/Account.cpp.o 
/usr/bin/c++  -DG_LOG_DOMAIN=\"gnc.engine\" -DHAVE_CONFIG_H -DHAVE_GUILE22 -D_GNU_SOURCE -Dgncmod_engine_EXPORTS -Ilibgnucash/engine -I/home/chris/sources/gnucash/maint/libgnucash/engine -Icommon -I/home/chris/sources/gnucash/maint/common -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/home/chris/sources/gnucash/maint/borrowed/libc -Ilibgnucash/core-utils -I/home/chris/sources/gnucash/maint/libgnucash/core-utils -I/usr/include/guile/2.2 -I/home/chris/sources/gnucash/maint/libgnucash/gnc-module -Wno-error=deprecated-declarations  -Werror -Wall -Wmissing-declarations -Wno-unused -Wno-error=parentheses -std=gnu++11  -g -fPIC -MD -MT libgnucash/engine/CMakeFiles/gncmod-engine.dir/Account.cpp.o -MF libgnucash/engine/CMakeFiles/gncmod-engine.dir/Account.cpp.o.d -o libgnucash/engine/CMakeFiles/gncmod-engine.dir/Account.cpp.o -c /home/chris/sources/gnucash/maint/libgnucash/engine/Account.cpp
/home/chris/sources/gnucash/maint/libgnucash/engine/Account.cpp: In function ‘GList* xaccAccountGetReconcileStatementDates(Account*)’:
/home/chris/sources/gnucash/maint/libgnucash/engine/Account.cpp:4426:22: error: request for member ‘get_keys’ in ‘slot’, which is of pointer type ‘KvpValueImpl*’ (maybe you meant to use ‘->’ ?)
 4426 |     auto keys = slot.get_keys ();
      |                      ^~~~~~~~
ninja: build stopped: subcommand failed.
```

Reply:
Why not? The error message tells you what to change.

Reply:
"Fixed" by exposing a private `kvp-frame.hpp` function.

Reply:
As for never being able to write C++ like that, why not? You probably write Python that's pretty similar. The `for (a : b)` is like Python's `foreach` (and there are folks on the C++ committee who hope to get something like Python list comprehensions in the next version, made possible by C++20's addition of coroutines).

The difference between `.` and `->` is the same as C: If you have `struct` you address its elements with `struct.a` and if you have a `struct*` you use `struct->a`. What's a little different is that C++ has a kind of non-nullable pointer called a reference and you use a `.` with those. I missed that KvpFrame::get_slot() returns a pointer instead of a reference.

BTW, slot.get_keys() can't return `nullptr` so there's no need to check it. It might return an empty `vector` but in that case the `for` loop won't do anything.